### ISSUE TITLE
davs network resource cannot handle dollar in password and does not support letsencrypt certificates

### ISSUE THREAD

**Issue by jdevrie** [xbmc/xbmc#19457]
### Describe the bug
It has taken me some time to find this out. 
On android on my Galaxy S10 I tried to connect with kodi to my synology webdav server using ssl.
I first had to add the letsencrypt CA certifcate authority certificate to the cacert.pem file to get a davs connection to work.
Then I still couldn't retrieve the directory listing. I got a 401 error while with exactly the same user and password in winscp it worked like a charm. 
Then I remembered having trouble with dollar signs in passwords in some packages because a dollar sign marks a variable in some programming languages (e.g PHP). 
So I changed the password of the user to a pw without a dollar and then it worked.
So it seemed like kodi is not handling the dollar sign properly in the webdavs url. 
As far as I knew a dollar in a url should be %24 but I saw something completely different in the URL it created. 
It changed the dollar sign into %c2%a7.
When I manualy changed that into %24 it worked just fine. So the editor of the network resource makes a mistake changing a dollar into %c2%a7 instead of %24.
It was quite a hassle to work this out.
Please add letsencrypt as a CA to kodi and fix this bug with the URL encoding of a dollar sign.

## Expected Behavior
Lets encrypt certificates are accepted by default.
Change a dollar sign into %24 in a webdav url

## Actual Behavior
Lets encrypt certificates are rejected because the CA is not in cacert.pem.
Change a dollar sign into %c2%a7 in a webdav url

## Possible Fix
You know that best

### To Reproduce
Read my story above

### Debuglog
NA

### Screenshots 
NA

## Additional context or screenshots (if appropriate)
NA

### Your Environment
Used Operating system: android and Synology DSM 7

 - [ ] Android version 11

 - Operating system version/name:
 - Kodi version: 19.0.0

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by lrusak** [comment]
You can add your own cert bundle

Ref: https://github.com/xbmc/xbmc/pull/19030

**Comment by jdevrie** [comment]
> You can add your own cert bundle
> 
> Ref: #19030

lrusak,
if you would have read the text properly you would have seen that I already did that. 
I think however that a CA like letsencrypt should be in the standard list provided by kodi since there are millions of NAS devices out there that use the lets encrypt CA.

**Comment by wsnipex** [comment]
our cacerts does contain the correct root CA for let's encrypt:
>Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            44:af:b0:80:d6:a3:27:ba:89:30:39:86:2e:f8:40:6b
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: O = Digital Signature Trust Co., CN = DST Root CA X3
        Validity
            Not Before: Sep 30 21:12:19 2000 GMT
            Not After : Sep 30 14:01:15 2021 GMT
        Subject: O = Digital Signature Trust Co., CN = DST Root CA X3


If you have issues, then you're servers cert chain might be missing the intermediate CA

**Comment by jdevrie** [comment]
@wsnipex
this is on android so it uses the cacert.pem of kodi.
,
You are right but the signing of leaf certificates is done with this certificate of LETS encrypt:
[https://crt.sh/?q=183267](https://crt.sh/?q=183267) 

So by default all the Synology NAS devices, and probably a lot of other brands too, are not accepted by kodi.
I was wondering why you do not use the default list of CA certificates on android.
Other android packages like filemanager+ use this and they have no problem with davs to my Synology NAS.

By the way the problem with the dollar sign is more of a nucance :-). This problem with the dollar sign only appears for davs
 not for dav. For dav the dollar sign is properly translated to %24. It happens when you use the "Add network resource" screen and klik ok. if you then edit the full url you see the %c2%a7 which you (luckily) can change manually into %24 and bypass this problem.

**Comment by wsnipex** [comment]
the cert you linked has expired in 2013...
>Version: 3 (0x2)
        Serial Number:
            74:38:78:f4:98:f8:a9:96:ab:c6:2b:eb:2e:90:6b:2a
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: (CA ID: 34)
            commonName                = VeriSign Class 3 International Server CA - G3
            organizationalUnitName    = Terms of use at https://www.verisign.com/rpa (c)10
            organizationalUnitName    = VeriSign Trust Network
            organizationName          = VeriSign, Inc.
            countryName               = US
        Validity (Expired)
            Not Before: May  7 00:00:00 2011 GMT
            Not After : Jun 20 23:59:59 2013 GMT

**Comment by jdevrie** [comment]
@wsnipex 
sorry I went a bit rougue while clicking on that site :-).
The proper root certificate for Letsencrypt is:

[https://crt.sh/?id=9314791](https://crt.sh/?id=9314791)

I put that one in the cacert.pem and it works with my letsencrypt certificates issued for my synology NAS.

**Comment by wsnipex** [comment]
again, that CA is in our cacerts.pem:
$ grep -i ISRG system/certs/cacert.pem 
ISRG Root X1

**Comment by jdevrie** [comment]
@wsnipex 
I know but it doesn't work with that certificate. 
I just found out that kodi does not  go from storage when I exit it. 
Only when I make android remove it form storage its gone.
I only get this to work when I put the intermediate certicate in the cacert.pem.
It looks like kodi does not drill up the hierarchy like other programs like winscp does.
All other webdav programs like winscp, mountainduck, filemanager+ can all succefully access the davs resource on my synology nas. The only one that can't is kodi unless I add something to the cacert.pem.

**Comment by wsnipex** [comment]
just provide the debug log that the issue template requires.
While you are at it, enable curl component logging

**Comment by jdevrie** [comment]
Hereby the debug log:
[kodi.log](https://github.com/xbmc/xbmc/files/6183133/kodi.log)

Here are the specific lines for the failed request:

2021-03-22 15:48:19.938 T:26220   DEBUG <general>: CurlFile::Open(0x731f198e28) https://admin.jdvl.nl:51006/Video/
2021-03-22 15:48:19.938 T:26220   DEBUG <general>: Curl::Debug - TEXT: RESOLVE admin.jdvl.nl:51006 is - old addresses discarded!
2021-03-22 15:48:19.938 T:26220   DEBUG <general>: Curl::Debug - TEXT: Added admin.jdvl.nl:51006:178.85.193.109 to DNS cache
2021-03-22 15:48:19.938 T:26220   DEBUG <general>: Curl::Debug - TEXT: Hostname admin.jdvl.nl was found in DNS cache
2021-03-22 15:48:19.939 T:26220   DEBUG <general>: Curl::Debug - TEXT:   Trying 178.85.193.109:51006...
2021-03-22 15:48:19.939 T:26220   DEBUG <general>: Curl::Debug - TEXT: TCP_NODELAY set
2021-03-22 15:48:19.978 T:26220   DEBUG <general>: Curl::Debug - TEXT: Connected to admin.jdvl.nl (178.85.193.109) port 51006 (#1)
2021-03-22 15:48:19.978 T:26220   DEBUG <general>: Curl::Debug - TEXT: ALPN, offering h2
2021-03-22 15:48:19.979 T:26220   DEBUG <general>: Curl::Debug - TEXT: ALPN, offering http/1.1
2021-03-22 15:48:19.991 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_OUT: 
2021-03-22 15:48:19.991 T:26220   DEBUG <general>: Curl::Debug - TEXT: TLSv1.2 (OUT), TLS handshake, Client hello (1):
2021-03-22 15:48:19.991 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_OUT: 
2021-03-22 15:48:19.991 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_OUT: À
2021-03-22 15:48:19.991 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_OUT: admin.jdvl.nl
2021-03-22 15:48:19.991 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_OUT: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Skipped 2 duplicate messages..
2021-03-22 15:48:20.015 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - TEXT: TLSv1.2 (IN), TLS handshake, Server hello (2):
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: -
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - TEXT: TLSv1.2 (IN), TLS handshake, Certificate (11):
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 	*H÷
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: Let's Encrypt10	UR30
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 210317011751Z
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 210615011751Z010U
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: admin.jdvl.nl0"0
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 	*H÷
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: {)y䵬ùVÿnrDz¯°콍蚍ýܗ񬆭ʝOù¨.񁉆K
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: άÁT桝©M>ªE	µ[3´ȐdꃹҌէL|?GO!򶹱ٖҐ҇Á,볚ģaƚ§´°ԡi9}ª»фEÀ`ł1%°x²<嫊G
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: ÿZlIۥ4G,
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: admin.jdvl.nl0LU E0C0g07+ߓ0(0&+http://cps.letsencrypt.org0
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: +ֹ􄁱
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: U맲|¿@񇅃WC0¯r퇘 G՞󚩊Ĥա·șΈ¨ÿö!攜󥊲021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: ¡À
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 爩 ±w4ǈsqQտ5$Ɋ/ҭ0ÿ­ ~@:xř1û¿{'!­-㺞 ¢YõJ~§峰
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 	*H÷
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: 
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: o4¥=`y��ꘀ0ʘ$3"¢ûی݁쒯ƫDFPƌp,
2021-03-22 15:48:20.016 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_IN: _ö#Jdt£Vº;ýT緗BS⹱��z2f²񌈵Y:ÿv«+o۽²W¯ÿ®ʸ䲁N;ݚ^Áy²±}eF¢nF𰣚t¬AbW®o䏨׍:܎\�³T¯mhփ,k¶°
2021-03-22 15:48:20.017 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_OUT: 
2021-03-22 15:48:20.017 T:26220   DEBUG <general>: Curl::Debug - TEXT: TLSv1.2 (OUT), TLS alert, unknown CA (560):
2021-03-22 15:48:20.017 T:26220   DEBUG <general>: Curl::Debug - SSL_DATA_OUT: 0
2021-03-22 15:48:20.017 T:26220   DEBUG <general>: Curl::Debug - TEXT: SSL certificate problem: unable to get local issuer certificate
2021-03-22 15:48:20.017 T:26220   DEBUG <general>: Curl::Debug - TEXT: Closing connection 1
2021-03-22 15:48:20.020 T:26220   ERROR <general>: CCurlFile::FillBuffer - Failed: SSL peer certificate or SSH remote key was not OK(60)
2021-03-22 15:48:20.020 T:26220   ERROR <general>: CCurlFile::Open failed with code 0 for https://admin.jdvl.nl:51006/Video/:
                                                   
2021-03-22 15:48:20.020 T:26220   ERROR <general>: GetDirectory - Unable to get dav directory (davs://USERNAME:PASSWORD@admin.jdvl.nl:51006/Video/)
2021-03-22 15:48:20.020 T:26220   ERROR <general>: GetDirectory - Error getting davs://USERNAME:PASSWORD@admin.jdvl.nl:51006/Video/

**Comment by wsnipex** [comment]
I can't open that log, but it doesn't matter. I already told you that it's likely an issue with your certificate chain server side...
$ openssl s_client -showcerts -connect admin.jdvl.nl:51006
CONNECTED(00000003)
depth=0 CN = admin.jdvl.nl
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = admin.jdvl.nl
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:CN = admin.jdvl.nl
   i:C = US, O = Let's Encrypt, CN = R3


As you can see, there is no intermediate cert in your chain, so the server config is broken.

**Comment by jdevrie** [comment]
Hereby my cacert.pem that works: 
[cacert.zip](https://github.com/xbmc/xbmc/files/6183227/cacert.zip)

**Comment by jdevrie** [comment]
If that is the case then why can all other programs connect to the same server with the same userid and password without problems and kodi can't.
That is  not logical. There must be something those programs are doing that kodi is not.

**Comment by wsnipex** [comment]
since this is perfectly reproducible with plain openssl, it has absolutely nothing to do with kodi.
Browsers have their own trust stores and they seem to include the "issuer=C = US, O = Let's Encrypt, CN = R3" intermediate CA that your server fails to include in the chain.
That might be convenient for you, but it won't work everywhere. The correct fix is to deliver that in the cert chain.

**Comment by jdevrie** [comment]
So the entire software world is clever enough to handle this situation properly except for open ssl.
Very strange.

**Comment by jdevrie** [comment]
By the way forget about the dollar sign replacement error. 
I found out that I had mistaken another sign (which I don't even have on my windows keyboard) for the dollar sign on my phone keyboard.

**Comment by jdevrie** [comment]
@wsnipex,
Hi Wolfgang,
you where right about the certificate chain.  I first looked at what webdav server is used on the synology server and that was apache. I looked around at the internet for how this is supposed to be configured. and found that the SSLCertificateChainFile keyword was to be used. I looked through the configuration files on my synology NAS as root and found the config  file where the SSLCertificateFile and SSLCertificateKeyFile keywords are located. 
There was no SSLCertificateChainFile  keyword there while there was actually a chainfile in the same direcotory as the SSLCertificateFile  and SSLCertificateKeyFile. So I added the SSLCertificateChainFile  keyword for that file to the configuration and now the openssl connections work without providing the intermediate certificate in the cacert.pem.

I learned a lot again.
Thanks for the help.
Regards,
Jan
